File: app/server/pkg/skat/rank.go
// Copyright 2025 Marcel Joachim Kloubert (https://marcel.coffee)
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package skat

import "fmt"

// Rank represents a card rank in Skat.
type Rank int

const (
	// Seven - lowest rank, 0 points
	Seven Rank = iota
	// Eight - 0 points
	Eight
	// Nine - 0 points
	Nine
	// Queen (Dame) - 3 points
	Queen
	// King (König) - 4 points
	King
	// Ten - 10 points
	Ten
	// Ace (Ass) - 11 points
	Ace
	// Jack (Bube) - 2 points, always trump in suit/grand games
	Jack
)

// AllRanks contains all ranks in standard order (lowest to highest for non-trump).
var AllRanks = []Rank{Seven, Eight, Nine, Queen, King, Ten, Ace, Jack}

// AllRanksWithoutJack contains ranks without Jack (for suit ordering).
var AllRanksWithoutJack = []Rank{Seven, Eight, Nine, Queen, King, Ten, Ace}

// NullRanks contains ranks in Null game order (A > K > Q > J > 10 > 9 > 8 > 7).
var NullRanks = []Rank{Seven, Eight, Nine, Ten, Jack, Queen, King, Ace}

// String returns the English name of the rank.
func (r Rank) String() string {
	switch r {
	case Seven:
		return "Seven"
	case Eight:
		return "Eight"
	case Nine:
		return "Nine"
	case Queen:
		return "Queen"
	case King:
		return "King"
	case Ten:
		return "Ten"
	case Ace:
		return "Ace"
	case Jack:
		return "Jack"
	default:
		return fmt.Sprintf("Rank(%d)", r)
	}
}

// GermanName returns the German name of the rank.
func (r Rank) GermanName() string {
	switch r {
	case Seven:
		return "Sieben"
	case Eight:
		return "Acht"
	case Nine:
		return "Neun"
	case Queen:
		return "Dame"
	case King:
		return "König"
	case Ten:
		return "Zehn"
	case Ace:
		return "Ass"
	case Jack:
		return "Bube"
	default:
		return fmt.Sprintf("Rang(%d)", r)
	}
}

// Code returns the ISS protocol code for the rank.
func (r Rank) Code() string {
	switch r {
	case Seven:
		return "7"
	case Eight:
		return "8"
	case Nine:
		return "9"
	case Queen:
		return "Q"
	case King:
		return "K"
	case Ten:
		return "T"
	case Ace:
		return "A"
	case Jack:
		return "J"
	default:
		return "?"
	}
}

// Points returns the point value of the rank.
func (r Rank) Points() int {
	switch r {
	case Ace:
		return 11
	case Ten:
		return 10
	case King:
		return 4
	case Queen:
		return 3
	case Jack:
		return 2
	case Nine, Eight, Seven:
		return 0
	default:
		return 0
	}
}

// RankFromCode parses a rank from its ISS protocol code.
func RankFromCode(code string) (Rank, error) {
	switch code {
	case "7":
		return Seven, nil
	case "8":
		return Eight, nil
	case "9":
		return Nine, nil
	case "Q":
		return Queen, nil
	case "K":
		return King, nil
	case "T":
		return Ten, nil
	case "A":
		return Ace, nil
	case "J":
		return Jack, nil
	default:
		return 0, fmt.Errorf("invalid rank code: %s", code)
	}
}

File: app/src/shared/rank.ts
// Copyright 2025 Marcel Joachim Kloubert (https://marcel.coffee)
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/**
 * Rank represents a card rank in Skat.
 */
export enum Rank {
  /** Seven - lowest rank, 0 points */
  Seven = 0,
  /** Eight - 0 points */
  Eight = 1,
  /** Nine - 0 points */
  Nine = 2,
  /** Queen (Dame) - 3 points */
  Queen = 3,
  /** King (König) - 4 points */
  King = 4,
  /** Ten - 10 points */
  Ten = 5,
  /** Ace (Ass) - 11 points */
  Ace = 6,
  /** Jack (Bube) - 2 points, always trump in suit/grand games */
  Jack = 7,
}

/** All ranks in standard order (lowest to highest for non-trump). */
export const ALL_RANKS: Rank[] = [
  Rank.Seven,
  Rank.Eight,
  Rank.Nine,
  Rank.Queen,
  Rank.King,
  Rank.Ten,
  Rank.Ace,
  Rank.Jack,
];

/** Ranks without Jack (for suit ordering). */
export const RANKS_WITHOUT_JACK: Rank[] = [
  Rank.Seven,
  Rank.Eight,
  Rank.Nine,
  Rank.Queen,
  Rank.King,
  Rank.Ten,
  Rank.Ace,
];

/** Ranks in Null game order (A > K > Q > J > 10 > 9 > 8 > 7). */
export const NULL_RANKS: Rank[] = [
  Rank.Seven,
  Rank.Eight,
  Rank.Nine,
  Rank.Ten,
  Rank.Jack,
  Rank.Queen,
  Rank.King,
  Rank.Ace,
];

/** Rank names in English. */
export const RANK_NAMES: Record<Rank, string> = {
  [Rank.Seven]: "Seven",
  [Rank.Eight]: "Eight",
  [Rank.Nine]: "Nine",
  [Rank.Queen]: "Queen",
  [Rank.King]: "King",
  [Rank.Ten]: "Ten",
  [Rank.Ace]: "Ace",
  [Rank.Jack]: "Jack",
};

/** Rank names in German. */
export const RANK_GERMAN_NAMES: Record<Rank, string> = {
  [Rank.Seven]: "Sieben",
  [Rank.Eight]: "Acht",
  [Rank.Nine]: "Neun",
  [Rank.Queen]: "Dame",
  [Rank.King]: "König",
  [Rank.Ten]: "Zehn",
  [Rank.Ace]: "Ass",
  [Rank.Jack]: "Bube",
};

/** ISS protocol codes for ranks. */
export const RANK_CODES: Record<Rank, string> = {
  [Rank.Seven]: "7",
  [Rank.Eight]: "8",
  [Rank.Nine]: "9",
  [Rank.Queen]: "Q",
  [Rank.King]: "K",
  [Rank.Ten]: "T",
  [Rank.Ace]: "A",
  [Rank.Jack]: "J",
};

/** Point values for ranks. */
export const RANK_POINTS: Record<Rank, number> = {
  [Rank.Seven]: 0,
  [Rank.Eight]: 0,
  [Rank.Nine]: 0,
  [Rank.Queen]: 3,
  [Rank.King]: 4,
  [Rank.Ten]: 10,
  [Rank.Ace]: 11,
  [Rank.Jack]: 2,
};

/**
 * Returns the English name of a rank.
 */
export function getRankName(rank: Rank): string {
  return RANK_NAMES[rank];
}

/**
 * Returns the German name of a rank.
 */
export function getRankGermanName(rank: Rank): string {
  return RANK_GERMAN_NAMES[rank];
}

/**
 * Returns the ISS protocol code for a rank.
 */
export function getRankCode(rank: Rank): string {
  return RANK_CODES[rank];
}

/**
 * Returns the point value of a rank.
 */
export function getRankPoints(rank: Rank): number {
  return RANK_POINTS[rank];
}

/**
 * Parses a rank from its ISS protocol code.
 */
export function rankFromCode(code: string): Rank | null {
  switch (code) {
    case "7":
      return Rank.Seven;
    case "8":
      return Rank.Eight;
    case "9":
      return Rank.Nine;
    case "Q":
      return Rank.Queen;
    case "K":
      return Rank.King;
    case "T":
      return Rank.Ten;
    case "A":
      return Rank.Ace;
    case "J":
      return Rank.Jack;
    default:
      return null;
  }
}
